### PR TITLE
Add `enableUnixSockets` option

### PR DESCRIPTION
**Pull request by TommyDew42**
### What is the PR for?

For issue #2046, we add an option to control if unix domain socket is supported:

```js
// enableUnixSocket is default to false before Got v13
const gotUnixSocketEnabled = got.extend({enableUnixSocket: true});
```

### How is the PR tested?

Unit tests.

#### Checklist

- [x] I have read the documentation.
- [x] I have included a pull request description of my changes.
- [x] I have included some tests.
- [x] If it's a new feature, I have included documentation updates in both the README and the types.

### CODE REVIEW

**Comment by sindresorhus** [comment]
Thanks for working on this. You also need to update the documentation.

https://github.com/sindresorhus/got/blob/main/documentation/tips.md#unix-domain-sockets can be moved to the options docs for this option.

File: documentation/2-options.md [hunk 1]
```diff
@@ -918,6 +918,28 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
 
 For example, when sending a `POST` request and receiving a `302`, it will resend the body to the new location using the same HTTP method (`POST` in this case). To rewrite the request as `GET`, set this option to `true`.
 
+### `enableUnixSocket`
+**Type: `boolean`**\
+**Default: `false`**
+
+Set to `true` to enable unix socket domain.
```
Comment (sindresorhus) [inline documentation/2-options.md:925]: This sentence is moot.

File: documentation/2-options.md [hunk 2]
```diff
@@ -918,6 +918,28 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
 
 For example, when sending a `POST` request and receiving a `302`, it will resend the body to the new location using the same HTTP method (`POST` in this case). To rewrite the request as `GET`, set this option to `true`.
 
+### `enableUnixSocket`
```
Comment (sindresorhus) [inline documentation/2-options.md:921]:
```suggestion
### `enableUnixSocket`

```

File: documentation/2-options.md [hunk 3]
```diff
@@ -918,6 +918,28 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
...
+**Type: `boolean`**\
+**Default: `false`**
+
+Set to `true` to enable unix socket domain.
+
+When it is enabled, requests can also be sent via [UNIX Domain Sockets](https://serverfault.com/questions/124517/what-is-the-difference-between-unix-sockets-and-tcp-ip-sockets).\
```
Comment (sindresorhus) [inline documentation/2-options.md:927]: Inconsistent pluralization of "socket".

Same in other places too.

File: documentation/2-options.md [hunk 4]
```diff
@@ -918,6 +918,28 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
...
+- `SOCKET` - Absolute path to a unix domain socket, for example: `/var/run/docker.sock`
+- `PATH` - Request path, for example: `/v2/keys`
+
+```js
+import got from 'got';
+
+await got('http://unix:/var/run/docker.sock:/containers/json');
+
+// Or without protocol (HTTP by default)
+await got('unix:/var/run/docker.sock:/containers/json');
```
Comment (sindresorhus) [inline documentation/2-options.md:940]: Missing the actual option.

File: documentation/tips.md [hunk 5]
```diff
@@ -89,24 +89,9 @@ for await (const commitData of pagination) {
...
-
-```js
-import got from 'got';
-
-await got('http://unix:/var/run/docker.sock:/containers/json');
-
-// Or without protocol (HTTP by default)
-await got('unix:/var/run/docker.sock:/containers/json');
-```
+See option [enableUnixSocket](./2-options.md#enableunixsocket) for more.
```
Comment (sindresorhus) [inline documentation/tips.md:94]:
```suggestion
See the [`enableUnixSocket` option](./2-options.md#enableunixsocket).
```

File: documentation/tips.md [hunk 6]
```diff
@@ -89,24 +89,9 @@ for await (const commitData of pagination) {
 }
 ```
 
-<a name="unix"></a>
```
Comment (sindresorhus) [inline documentation/tips.md:92]: Keep this

File: documentation/2-options.md [hunk 7]
```diff
@@ -918,6 +918,28 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
 
 For example, when sending a `POST` request and receiving a `302`, it will resend the body to the new location using the same HTTP method (`POST` in this case). To rewrite the request as `GET`, set this option to `true`.
 
+### `enableUnixSocket`
+**Type: `boolean`**\
+**Default: `false`**
+
+Set to `true` to enable unix socket domain.
+
```
Comment (sindresorhus) [inline documentation/2-options.md:926]: There should be a note to about recommending not to enable this option if you accept untrusted user input for the URL.

File: test/redirects.ts [hunk 8]
```diff
@@ -42,10 +42,31 @@ const unixHostname: Handler = (_request, response) => {
...
+		instanceOf: RequestError,
+	});
+
+	await t.throwsAsync(gotUnixSocketEnabled('hostname'), {
+		message: 'Cannot redirect to UNIX socket',
+		instanceOf: RequestError,
+	});
+});
+
+test('cannot redirect to unix protocol when unix socket is not enabled', withServer, async (t, server, got) => {
```
Comment (sindresorhus) [inline test/redirects.ts:64]:
```suggestion
test('cannot redirect to UNIX protocol when UNIX socket is not enabled', withServer, async (t, server, got) => {
```

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: documentation/2-options.md (hunk 3 above)
Comment (TommyDew42) [reply documentation/2-options.md:927]: What does it mean?
Do we wanna rename the option from `enableUnixSocket` to `enableUnixSockets`?

Comment (sindresorhus) [reply documentation/2-options.md:928]: > Do we wanna rename the option from enableUnixSocket to enableUnixSockets?

Yes, and also mention of UNIX Sockets should use the plural version.

Comment (TommyDew42) [reply documentation/2-options.md:928]: Changed in 7ac2d9cf8fbf17d27677434842a3585f16a81e49

**Comment by TommyDew42** [comment]
@sindresorhus Do you think rerunning the test will work? Don’t think the extra if-statement made us exceed the limit.

File: documentation/2-options.md [hunk 9]
```diff
@@ -918,6 +918,31 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
 
 For example, when sending a `POST` request and receiving a `302`, it will resend the body to the new location using the same HTTP method (`POST` in this case). To rewrite the request as `GET`, set this option to `true`.
 
+### `enableUnixSockets`
+
+**Type: `boolean`**\
+**Default: `false`**
+
+Enable it with care if you accept untrusted user input for the URL.
```
Comment (sindresorhus) [inline documentation/2-options.md:926]: This should be below the description of the option.

File: documentation/2-options.md [hunk 10]
```diff
@@ -918,6 +918,31 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
 
 For example, when sending a `POST` request and receiving a `302`, it will resend the body to the new location using the same HTTP method (`POST` in this case). To rewrite the request as `GET`, set this option to `true`.
 
+### `enableUnixSockets`
+
+**Type: `boolean`**\
+**Default: `false`**
```
Comment (sindresorhus) [inline documentation/2-options.md:924]: The option should be `true` by default as making it `false` is a breaking change.

File: documentation/2-options.md [hunk 11]
```diff
@@ -918,6 +918,31 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
...
+
+Enable it with care if you accept untrusted user input for the URL.
+
+When it is enabled, requests can also be sent via [UNIX Domain Sockets](https://serverfault.com/questions/124517/what-is-the-difference-between-unix-sockets-and-tcp-ip-sockets).\
+Use the following URL scheme: `PROTOCOL://unix:SOCKET:PATH`
+
+- `PROTOCOL` - `http` or `https`
+- `SOCKET` - Absolute path to a unix domain socket, for example: `/var/run/docker.sock`
```
Comment (sindresorhus) [inline documentation/2-options.md:932]:
```suggestion
- `SOCKET` - Absolute path to a UNIX domain socket, for example: `/var/run/docker.sock`
```

File: documentation/2-options.md (hunk 9 above)
Comment (sindresorhus) [inline documentation/2-options.md:926]: > And make sure you do your own URL sanitizing.

File: test/redirects.ts [hunk 12]
```diff
@@ -42,10 +42,31 @@ const unixHostname: Handler = (_request, response) => {
...
 };
 
-test('cannot redirect to unix protocol', withServer, async (t, server, got) => {
+test('cannot redirect to UNIX protocol when UNIX sockets are enabled', withServer, async (t, server, got) => {
 	server.get('/protocol', unixProtocol);
 	server.get('/hostname', unixHostname);
 
+	const gotUnixSocketsEnabled = got.extend({enableUnixSockets: true});
+
+	t.assert(gotUnixSocketsEnabled.defaults.options.enableUnixSockets);
```
Comment (sindresorhus) [inline test/redirects.ts:51]:
```suggestion
	t.true(gotUnixSocketsEnabled.defaults.options.enableUnixSockets);
```

File: test/unix-socket.ts [hunk 13]
```diff
@@ -50,22 +52,48 @@ if (process.platform !== 'win32') {
...
 	test('redirects work', withSocketServer, async (t, server) => {
 		server.on('/', redirectHandler);
 		server.on('/foo', okHandler);
 
 		const url = format('http://unix:%s:%s', server.socketPath, '/');
-		t.is((await got(url)).body, 'ok');
+		t.is((await gotUnixSocketsEnabled(url)).body, 'ok');
+	});
+
+	test('unix: fails when unix sockets are not enabled', async t => {
```
Comment (sindresorhus) [inline test/unix-socket.ts:74]:
```suggestion
	test('`unix:` fails when UNIX sockets are not enabled', async t => {
```

File: test/unix-socket.ts [hunk 14]
```diff
@@ -50,22 +52,48 @@ if (process.platform !== 'win32') {
...
+		try {
+			t.assert(!got.defaults.options.enableUnixSockets);
```
Comment (sindresorhus) [inline test/unix-socket.ts:76]:
```suggestion
			t.false(got.defaults.options.enableUnixSockets);
```

File: test/unix-socket.ts [hunk 15]
```diff
@@ -50,22 +52,48 @@ if (process.platform !== 'win32') {
...
+			await got('unix:');
+		} catch (error: any) {
```
Comment (sindresorhus) [inline test/unix-socket.ts:78]: Use `t.throwsAsync`

Comment (sindresorhus) [reply test/unix-socket.ts:78]: Same with the below test.

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: documentation/2-options.md [hunk 16]
```diff
@@ -918,6 +918,35 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
...
 
+### `enableUnixSockets`
+
+**Type: `boolean`**\
+**Default: `true`**
+
+When it is enabled, requests can also be sent via [UNIX Domain Sockets](https://serverfault.com/questions/124517/what-is-the-difference-between-unix-sockets-and-tcp-ip-sockets).\
+Use the following URL scheme: `PROTOCOL://unix:SOCKET:PATH`
+
+By default it is enabled, so make sure you do your own URL sanitizing if you accept untrusted user input for the URL.
```
Comment (TommyDew42) [inline documentation/2-options.md:929]: Rewrite the reminder to highlight it is enabled by default.

File: test/unix-socket.ts [hunk 17]
```diff
@@ -68,4 +68,32 @@ if (process.platform !== 'win32') {
...
+
+	test('`http://unix:/` fails when UNIX sockets are not enabled', async t => {
+		const gotUnixSocketsDisabled = got.extend({enableUnixSockets: false});
+
+		t.false(gotUnixSocketsDisabled.defaults.options.enableUnixSockets);
+
+		try {
+			await got('http://unix:');
+		} catch (error: any) {
+			t.regex(error.code, /ENOTFOUND|EAI_AGAIN/);
```
Comment (TommyDew42) [inline test/unix-socket.ts:92]: We want to check with regex here. Seems `t.throwsAsync` has not supported regex in `code` yet.

File: documentation/2-options.md [hunk 18]
```diff
@@ -918,6 +918,35 @@ By default, requests will not use [method rewriting](https://datatracker.ietf.or
...
+await got('http://unix:/var/run/docker.sock:/containers/json');
+
+// Or without protocol (HTTP by default)
+await got('unix:/var/run/docker.sock:/containers/json');
+
+// Disable Unix sockets
+const gotUnixSocketsDisabled = got.extend({enableUnixSockets: false});
+
+// Error!
+gotUnixSocketsDisabled('http://unix:/var/run/docker.sock:/containers/json')
```
Comment (TommyDew42) [inline documentation/2-options.md:947]: Added example when `enableUnixSockets` is `false`

File: documentation/2-options.md (hunk 16 above)
Comment (szmarczak) [reply documentation/2-options.md:929]: Can you make this a note like this: https://github.com/sindresorhus/got/blob/main/documentation/2-options.md#note-3

File: source/core/options.ts [hunk 19]
```diff
@@ -1401,7 +1402,7 @@ export default class Options {
 		this._internals.url = url;
 		decodeURI(urlString);
 
-		if (url.protocol === 'unix:') {
+		if (this._internals.enableUnixSockets && url.protocol === 'unix:') {
```
Comment (szmarczak) [inline source/core/options.ts:1405]: Please revert this change. Instead check if this is not enabled inside this if https://github.com/sindresorhus/got/blob/d4c2913ec25968e8d99a4a24e589d3e741072369/source/core/options.ts#L1430 and throw.

File: source/core/options.ts [hunk 20]
```diff
@@ -1427,7 +1428,7 @@ export default class Options {
 			this._internals.searchParams = undefined;
 		}
 
-		if (url.hostname === 'unix') {
+		if (this._internals.enableUnixSockets && url.hostname === 'unix') {
```
Comment (szmarczak) [inline source/core/options.ts:1431]: Please revert this change.

File: test/unix-socket.ts (hunk 17 above)
Comment (szmarczak) [reply test/unix-socket.ts:92]: Why `EAI_AGAIN`?

Comment (sindresorhus) [reply test/unix-socket.ts:92]: `t.throwsAsync` returns the error too. So you can check the return value.

Comment (TommyDew42) [reply test/unix-socket.ts:92]: `EAI_AGAIN` is for ubuntu. The error code there is different from mac & win.

I log the error in a previous [test](https://github.com/sindresorhus/got/runs/6949849768?check_suite_focus=true#step:5:729).
Screenshot 2022-07-09 at 1 02 18 PM

File: source/core/options.ts [hunk 21]
```diff
@@ -1428,6 +1429,10 @@ export default class Options {
 		}
 
 		if (url.hostname === 'unix') {
+			if (!this._internals.enableUnixSockets) {
+				throw new Error('Using UNIX domain sockets but option `enableUnixSockets` is not enabled');
+			}
+
```
Comment (TommyDew42) [inline source/core/options.ts:1435]: Followed @szmarczak's suggestion to check if the option is enabled. Throw an error if not enabled.

File: test/unix-socket.ts [hunk 22]
```diff
@@ -68,4 +68,29 @@ if (process.platform !== 'win32') {
...
+
+		t.false(gotUnixSocketsDisabled.defaults.options.enableUnixSockets);
+
+		await t.throwsAsync(
+			gotUnixSocketsDisabled('http://unix:'),
+			{
+				message: 'Using UNIX domain sockets but option `enableUnixSockets` is not enabled',
+			},
+		);
+	});
```
Comment (TommyDew42) [inline test/unix-socket.ts:95]: Changed tests for the custom error.

**Review by sindresorhus — APPROVED** [review]

**Review by szmarczak — APPROVED** [review]
LGTM

**Comment by TommyDew42** [comment]
Seems the [release](https://nodejs.org/es/blog/release/v18.6.0/) of node v18.6.0 on 13/7 broke the tests. Is this [commit](https://github.com/nodejs/node/commit/4df96b501d) that failed the test?

**Comment by szmarczak** [comment]
Don't worry about that. We'll fix that in the `main` branch, that doesn't stop us from merging :) Can you please address https://github.com/sindresorhus/got/pull/2062#discussion_r925577160

File: test/unix-socket.ts [hunk 23]
```diff
@@ -68,4 +68,32 @@ if (process.platform !== 'win32') {
 		const url = format('http://unix:%s:%s', server.socketPath, '/');
 		t.is((await got(url)).body, 'ok');
 	});
+
+	test('`unix:` fails when UNIX sockets are not enabled', async t => {
+		const gotUnixSocketsDisabled = got.extend({enableUnixSockets: false});
+
+		t.false(gotUnixSocketsDisabled.defaults.options.enableUnixSockets);
```
Comment (szmarczak) [inline test/unix-socket.ts:75]: This is quite useless. You an just call `got('unix:', {enableUnixSockets: false})` which does the same.

File: test/unix-socket.ts (hunk 17 above)
Comment (szmarczak) [reply test/unix-socket.ts]: This seems incorrect. The error codes mean that the request actually went through (even though it wasn't unix, it looked up the `unix` hostname instead).

**Review by szmarczak — CHANGES_REQUESTED** [review]

**Comment by szmarczak** [comment]
Thanks @TommyDew42 ❤️
